### PR TITLE
Adds ansible for end-of-contest procedure.

### PR DESCRIPTION
**Pull request by tuupke**
The readme contains some explanation on how to use the playbook. Note, this is probably still quite limited, so check the results before relying on them!

### CODE REVIEW

File: provision-contest/ansible/EOC/README.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
+## DOMjudge EOC ansible helper
+This ansible playbook should automate must of the tasks DOMjude needs to perform for the End Of Contest procedures during the ICPC WF's. It was created for the WF 2021 in Dahka, which took place in 2022.
```
Comment (vmcj) [inline provision-contest/ansible/EOC/README.md:2]:
```suggestion
This ansible playbook should automate must of the tasks DOMjudge needs to perform for the End Of Contest procedures during the ICPC WF's. It was created for the WF 2021 in Dakha, which took place in 2022.
```

File: provision-contest/ansible/EOC/README.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+Two types of authentication are used and supported. Session based auth and basic auth. 
+ - Session based auth retrieves a CSRF token and fakes a login action.
+ - Basic auth headers are constructed and used. This is to fully mitigate [annoying basic-auth issues](https://askubuntu.com/questions/1070838/why-wget-does-not-use-username-and-password-in-url-first-time/1070847#1070847), even though the `force_basic_auth` option should already prevent these issues from arising.
+
+The full playbook can be run using the following command: `ansible-playbook main.yml --extra-vars "@facts.yml" --tags all`. See the setup section to see what the contents of facts.yml are supposed to be.
```
Comment (meisterT) [inline provision-contest/ansible/EOC/README.md:8]: Super-nit: move the command to an extra line you it's even easier to copy/ paste.

File: provision-contest/ansible/EOC/README.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
+## DOMjudge EOC ansible helper
+This ansible playbook should automate must of the tasks DOMjudge needs to perform for the End Of Contest procedures during the ICPC WF's. It was created for the WF 2021 in Dhaka, which took place in 2022.
```
Comment (vmcj) [inline provision-contest/ansible/EOC/README.md:2]:
```suggestion
This ansible playbook should automate most of the tasks DOMjudge needs to perform for the End Of Contest procedures during the ICPC WF's. It was created for the WF 2021 in Dhaka, which took place in 2022.
```

File: provision-contest/ansible/EOC/README.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+```ansible-playbook main.yml --extra-vars "@facts.yml" --tags all```
+
+### Steps automated.
+The following steps from the EOC 2022-Oct document have been automated:
+
+29. `Primary pushes results.tsv into git repo`. Tag for just this action: `results.tsv`. Note, a manual commit is still needed!
+32. `Compare scoreboard.json between primary and shadow`. Tag for just this action: `scoreboard`. Note, only if other (shadow or primary) credentials are provided.
+33. `Compare awards.json between primary and CDS`. Tag for just this action: `awards`. Note, only if CDS credentials are provided.
+36. `Verify that Primary and ... System Test`. Tag for just this action: `fetch`. Note, a manual commit is still needed and the `final_standings.html` is actually the static-scoreboard zip.
```
Comment (vmcj) [inline provision-contest/ansible/EOC/README.md:18]: DOMjudge has an actual final_standings.html which is probably expected to be used here?

File: provision-contest/ansible/EOC/main.yml [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+      ansible.builtin.set_fact:
+        cds_basic_auth: "Basic {{ (cds_username+':'+cds_password) | b64encode }}"
+      when: cds_available
+
+    - name: Set other basic auth fact
+      tags: all, compare, scoreboard, awards
+      ansible.builtin.set_fact:
+        other_basic_auth: "Basic {{ (other_username+':'+other_password) | b64encode }}"
+      when: other_available
+
```
Comment (vmcj) [inline provision-contest/ansible/EOC/main.yml:29]:
```suggestion

```

File: provision-contest/ansible/EOC/main.yml [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+          register: login_response
+
+        - name: Test succesfull login by going to a restricted page
+          ansible.builtin.uri:
+            url: "{{ dj_url }}/jury"
+            method: GET
+            headers:
+              cookie: "{{ login_response.set_cookie }}"
+
+
```
Comment (vmcj) [inline provision-contest/ansible/EOC/main.yml:66]:
```suggestion
```

File: provision-contest/ansible/EOC/README.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+   - groups
+   - judgements
+   - languages
+   - organizations
+   - runs
+   - scoreboard
+   - teams
+
+### Known limitations
+ - The checks that verify whether the jsons are the same are *very* simple and I (Mart) have not yet found a reasonable way of actually checking them in a nice manner from ansible. Them matching would be a bigger red-flag than them having a difference. Still, the playbook fails when a difference is detected! To aid with (manual) verification the results of fetching the jsons is stored in `/tmp/[obj].[sys].json` with `[object]` either "awards" or "scoreboard" and `[sys]` either "dj", "cds", or "other".
```
Comment (nickygerritsen) [inline provision-contest/ansible/EOC/README.md:79]: We should use the icpc tools contestutil for this

Comment (tuupke) [reply provision-contest/ansible/EOC/README.md:79]: If you have documentation for that for me. I'll make sure it works. :)

File: provision-contest/ansible/EOC/README.md (hunk 3 above)
Comment (tuupke) [reply provision-contest/ansible/EOC/README.md]: Fixed

File: provision-contest/ansible/EOC/main.yml (hunk 5 above)
Comment (tuupke) [reply provision-contest/ansible/EOC/main.yml:29]: Fixed

File: provision-contest/ansible/EOC/README.md (hunk 4 above)
Comment (tuupke) [reply provision-contest/ansible/EOC/README.md]: Fixed

File: provision-contest/ansible/EOC/README.md (hunk 7 above)
Comment (nickygerritsen) [reply provision-contest/ansible/EOC/README.md:79]:
```
./contestUtil.sh ScoreboardUtil ~/Desktop/dj.json ~/Desktop/cdsm.json # for scoreboard compare
```

For awards it doesn't exist yet... There is a way to compare the whole event feed.

@deboer-tim could we get an awards.json compare as a separate util?

File: provision-contest/ansible/EOC/main.yml (hunk 6 above)
Comment (tuupke) [reply provision-contest/ansible/EOC/main.yml]: Fixed

File: provision-contest/ansible/EOC/README.md (hunk 7 above)
Comment (tuupke) [reply provision-contest/ansible/EOC/README.md]: Implemented. awards.json still makes use of the borked jd method.

**Review by vmcj — APPROVED** [review]
As we want to test this I'll merge this.

@eldering we still need a make target for this, I'll make something but if you're faster when fixing that would be fine.

**Pull request merged**